xbmc/addons/PVRClient.cpp

```diff
@@ -1565,29 +1589,432 @@ void CPVRClient::OnSystemSleep(void)
+  CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, strLine1, strLine2, 5000, false);
+  CEventLog::GetInstance().Add(EventPtr(new CNotificationEvent(client->Name(), strLine1, client->Icon(), strLine2)));
+
+  CLog::Log(LOGDEBUG, "PVR - %s - recording %s on client '%s'. name='%s' filename='%s'",
+      __FUNCTION__, bOnOff ? "started" : "finished", client->Name().c_str(), strName, strFileName);
+}
+
+void CPVRClient::cb_trigger_channel_update(void *addonData)
+{
+  /* update the channels table in the next iteration of the pvrmanager's main loop */
+  CServiceBroker::GetPVRManager().TriggerChannelsUpdate();
+}
+
+void CPVRClient::cb_trigger_timer_update(void *addonData)
+{
+  /* update the timers table in the next iteration of the pvrmanager's main loop */
+  CServiceBroker::GetPVRManager().TriggerTimersUpdate();
+}
+
+void CPVRClient::cb_trigger_recording_update(void *addonData)
+{
+  /* update the recordings table in the next iteration of the pvrmanager's main loop */
+  CServiceBroker::GetPVRManager().TriggerRecordingsUpdate();
+}
+
+void CPVRClient::cb_trigger_channel_groups_update(void *addonData)
+{
+  /* update all channel groups in the next iteration of the pvrmanager's main loop */
+  CServiceBroker::GetPVRManager().TriggerChannelGroupsUpdate();
+}
+
+void CPVRClient::cb_trigger_epg_update(void *addonData, unsigned int iChannelUid)
+{
+  // get the client
+  CPVRClient *client = static_cast<CPVRClient*>(addonData);
+  if (!client)
+  {
+    CLog::Log(LOGERROR, "PVR - %s - invalid handler data", __FUNCTION__);
+    return;
+  }
+
+  CServiceBroker::GetPVRManager().EpgContainer().UpdateRequest(client->GetID(), iChannelUid);
+}
+
+void CPVRClient::cb_free_demux_packet(void *addonData, DemuxPacket* pPacket)
+{
+  CDVDDemuxUtils::FreeDemuxPacket(pPacket);
+}
+
+DemuxPacket* CPVRClient::cb_allocate_demux_packet(void *addonData, int iDataSize)
+{
+  return CDVDDemuxUtils::AllocateDemuxPacket(iDataSize);
+}
+
+void CPVRClient::cb_connection_state_change(void* addonData, const char* strConnectionString, PVR_CONNECTION_STATE newState, const char *strMessage)
+{
+  CPVRClient *client = static_cast<CPVRClient*>(addonData);
+  if (!client || !strConnectionString)
+  {
+    CLog::Log(LOGERROR, "PVR - %s - invalid handler data", __FUNCTION__);
+    return;
+  }
+
+  const PVR_CONNECTION_STATE prevState(client->GetConnectionState());
+  if (prevState == newState)
+    return;
+
+  CLog::Log(LOGDEBUG, "PVR - %s - state for connection '%s' on client '%s' changed from '%d' to '%d'", __FUNCTION__, strConnectionString, client->Name().c_str(), prevState, newState);
+
+  client->SetConnectionState(newState);
+
+  std::string msg;
+  if (strMessage != nullptr)
+    msg = strMessage;
+
+  CServiceBroker::GetPVRManager().ConnectionStateChange(client, std::string(strConnectionString), newState, msg);
+}
+
+typedef struct EpgEventStateChange
+{
+  int             iClientId;
+  unsigned int    iUniqueChannelId;
+  CPVREpgInfoTagPtr  event;
```

Comment:
Could you please fix the spaces? I suggest to put only one blank between type and variable name.